hazelcast/src/main/java/com/hazelcast/concurrent/flakeidgen/FlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.concurrent.flakeidgen;
+
+import com.hazelcast.core.FlakeIdGenerator;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.AbstractDistributedObject;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.util.Clock;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class FlakeIdGeneratorProxy
```

Comment:
Can you put some javadoc on this class. Specifically, I would like to see some short explanation of how IDs are generated and what's the binary layout. Alternatively, you can add this to the `newIdBase` method.